### PR TITLE
Transform System Rewrite/Update

### LINKED ISSUE #336: Update the Transform System
I wrote the transform system when specs was still a relatively newer ecs and ended up taking a lot of ideas from that and implementing most of them into specs in a more performant/space efficient way. Now that those components are in specs, it would be better to rewrite the transform system in that style.

- [ ] Replace boolean flags with `FlaggedStorage`
- [ ] Use bitset joining instead of an `Init` component.
- [ ] Rename `Child` back to `Parent`? (I was indifferent when I wrote it, but I feel that `Parent` is a better identifier since the entity is contains is the "parent" entity)

There are probably going to be some annoying collisions with how the old transform system worked when trying to port it over, so this might take longer than expected.

### PR DESCRIPTION
**Pull request by Aceeri**
Cleans up the transform system with some new specs features, adds unit tests.

Still plan on adding some tests using `quickcheck` for some more "fuzzy" testing to catch the odd edge cases.

Fixes #336

### CODE REVIEW

File: examples/05_assets/main.rs [hunk 1]
```diff
@@ -13,8 +13,8 @@ use amethyst::config::Config;
 use amethyst::ecs::World;
 use amethyst::ecs::input::InputBundle;
 use amethyst::ecs::rendering::{AmbientColor, Factory, LightComponent, MaterialComponent,
-                               MeshComponent, MeshContext, RenderBundle, TextureComponent,
-                               TextureContext};
+MeshComponent, MeshContext, RenderBundle, TextureComponent,
```
Comment (Rhuagh) [inline examples/05_assets/main.rs]: This looks like broken formatting, are you using the latest rustfmt-nightly? And running cargo +nightly fmt --all ?

File: src/ecs/transform/systems.rs [hunk 2]
```diff
@@ -209,4 +168,211 @@ mod tests {
...
+        let mut transform = LocalTransform::default();
+        transform.translation = [0.0, 0.0, 0.0];
+        transform.rotation = [1.0, 0.0, 0.0, 0.0];
+
+        let e1 = world.create_entity()
+            .with(transform)
+            .with(Transform::default())
+            .build();
+
+        dispatcher.dispatch(&mut world.res);
```
Comment (Rhuagh) [inline src/ecs/transform/systems.rs]: Why not just do `RunNow` on the system? Feels like unnecessary overhead to use a real `Dispatcher`.

Comment (Aceeri) [reply src/ecs/transform/systems.rs]: Was just easier to set up, overhead doesn't really matter that much in these tests anyways.

Comment (torkleyy) [reply src/ecs/transform/systems.rs]: How is it easier? It's more lines.

File: examples/05_assets/main.rs (hunk 1 above)
Comment (Aceeri) [reply examples/05_assets/main.rs]: I don't really know what went on here either, I ran the latest nightly but I'll try again this afternoon.

File: src/ecs/transform/systems.rs (hunk 2 above)
Comment (Aceeri) [reply src/ecs/transform/systems.rs]: Mainly because I know it by memory,.

Comment (Rhuagh) [reply src/ecs/transform/systems.rs]: `system.run_now(&mut world.res)` ;)

**Comment by Rhuagh** [comment]
How is this coming along? Any progress? Is it done or ?

**Comment by Aceeri** [comment]
@Rhuagh Got held up for a couple days this week, working on it this weekend.

**Comment by Rhuagh** [comment]
Cool

**Comment by Aceeri** [comment]
Been fuzzy testing and I think I've fixed all the issues so far, so we will probably be able to merge this by the end of the week.

**Comment by torkleyy** [comment]
Seems @nchashch wants to review this before merge.

**Comment by Aceeri** [comment]
Yep, just updating you all on it.

**Comment by Rhuagh** [comment]
Did you rebase against your local develop our the upstream develop?

**Comment by Xaeroxe** [comment]
It would probably be easier to rewrite this PR at this point.  Please re-open if you disagree with me @Aceeri

**Comment by Aceeri** [comment]
I disagree, the commits earlier are just an error in git and I just have to drop the commits due to duplicates.

**Comment by Aceeri** [comment]
Issue with commits has been fixed.

**Comment by Xaeroxe** [comment]
Do you still consider this to be WIP?

**Comment by Aceeri** [comment]
Not really, it could use some review though, entirely possible I've missed something.

**Comment by Rhuagh** [comment]
I'll take a look later today.

File: amethyst_core/src/transform/systems.rs [hunk 3]
```diff
@@ -169,17 +185,29 @@ impl<'a> System<'a> for TransformSystem {
             index += 1;
         }
 
-        self.dirty.clear();
+        println!("with parent updated: {:?}", parented);
```
Comment (Rhuagh) [inline amethyst_core/src/transform/systems.rs]: Debug prints should be removed?

File: amethyst_core/src/transform/systems.rs [hunk 4]
```diff
@@ -169,17 +185,29 @@ impl<'a> System<'a> for TransformSystem {
...
+        for bit in (&self.frame_init).iter() {
+            self.init.add(bit);
+        }
+        self.frame_init.clear();
         self.dead.clear();
-        self.swapped.clear();
     }
 }
 
 #[cfg(test)]
```
Comment (Rhuagh) [inline amethyst_core/src/transform/systems.rs]: Yay, tests!

File: amethyst_core/src/transform/systems.rs (hunk 3 above)
Comment (Aceeri) [reply amethyst_core/src/transform/systems.rs]: whoops

File: amethyst_core/src/transform/systems.rs [hunk 5]
```diff
@@ -31,87 +29,106 @@ impl TransformSystem {
...
-        }
-
-        // Deletes entities whose parents aren't alive.
-        for &(entity, _) in &self.sorted {
-            if let Some(child) = children.get(entity) {
-                if !entities.is_alive(child.parent()) || self.dead.contains(&child.parent()) {
-                    let _ = entities.delete(entity);
-                    self.dead.insert(entity);
+        {
+            for (entity, parent) in (&*entities, parents.open().1).join() {
```
Comment (torkleyy) [inline amethyst_core/src/transform/systems.rs]: We need `retain`!

Comment (torkleyy) [reply amethyst_core/src/transform/systems.rs]: But isn't this too heavy given that it would be very rare?

File: amethyst_core/src/transform/systems.rs [hunk 6]
```diff
@@ -31,87 +29,106 @@ impl TransformSystem {
...
+            let filter = locals.open().0 & parents.open().0 & !&self.init; // has a local, parent, and isn't initialized.
+            for (entity, _) in (&*entities, &filter).join() {
+                self.indices.insert(entity, self.sorted.len());
+                self.sorted.push(entity);
+                self.frame_init.add(entity.id());
+
+                initted.push(entity.id());
+            }
+
+            println!("Initialized: {:?}", initted);
```
Comment (torkleyy) [inline amethyst_core/src/transform/systems.rs]: Another debugging println

File: amethyst_core/src/transform/systems.rs [hunk 7]
```diff
@@ -31,87 +29,106 @@ impl TransformSystem {
...
-            {
-                if local.is_dirty() {
-                    self.dirty.insert(ent);
-                    let global = global_restrict.get_mut_unchecked(&mut global_entry);
-                    global.0 = local.matrix();
-                    local.flag(false);
+            for (_entity, local, global, _) in (&*entities, locals_flagged, &mut globals, !&parents).join() {
+                global.0 = local.matrix();
+                noparents.push(_entity.id());
+                #[cfg(debug_assertions)]
```
Comment (torkleyy) [inline amethyst_core/src/transform/systems.rs]: Should use `debug_assert` and `global.0.is_finite()`.

File: amethyst_core/src/transform/systems.rs [hunk 8]
```diff
@@ -169,17 +185,29 @@ impl<'a> System<'a> for TransformSystem {
...
         }
 
-        self.dirty.clear();
+        println!("with parent updated: {:?}", parented);
+        println!("no parent updated: {:?}", noparents);
+
+        (&mut locals).open().1.clear_flags();
+        (&mut parents).open().1.clear_flags();
+
+        for bit in (&self.frame_init).iter() {
```
Comment (torkleyy) [inline amethyst_core/src/transform/systems.rs]: No need for `.iter()` I think

File: amethyst_core/src/transform/components/parent.rs [hunk 9]
```diff
@@ -0,0 +1,13 @@
+
+use specs::{Component, DenseVecStorage, Entity, FlaggedStorage};
+
+/// Component for defining a parent entity.
```
Comment (torkleyy) [inline amethyst_core/src/transform/components/parent.rs]: This documentation should exactly document on which entity one has to place the component. (It's confusing whether this is a "has" relation or an "is parent of".

File: amethyst_core/src/transform/systems.rs (hunk 5 above)
Comment (Aceeri) [reply amethyst_core/src/transform/systems.rs]: It is only running on modified parents, so I don't think it should be that heavy.

Comment (torkleyy) [reply amethyst_core/src/transform/systems.rs]: I'm not sure yet if `FlaggedStorage` is user friendly enough tbh.

Comment (Aceeri) [reply amethyst_core/src/transform/systems.rs]: Yeah, the issue is you'd have to expose the underlying UnprotectedStorage methods onto Storage if you wanted it to be more straightforward.

File: amethyst_core/src/transform/components/parent.rs (hunk 9 above)
Comment (Rhuagh) [reply amethyst_core/src/transform/components/parent.rs]: Rename it to `HasParent` maybe ?

File: amethyst_core/src/transform/systems.rs (hunk 8 above)
Comment (Aceeri) [reply amethyst_core/src/transform/systems.rs]: BitSetLikes don't implement IntoIterator, but they probably should so: https://github.com/slide-rs/hibitset/pull/18

File: amethyst_core/src/transform/systems.rs [hunk 10]
```diff
@@ -120,46 +123,44 @@ impl<'a> System<'a> for TransformSystem {
...
-                            self.swapped.insert(entity);
 
                             // Swap took place, re-try this index.
                             continue;
                         }
                     }
 
-                    if local.is_dirty() || child.is_dirty() || self.dirty.contains(&child.parent())
-                    {
+                    // Kill the entity is the parent is dead.
```
Comment (torkleyy) [inline amethyst_core/src/transform/systems.rs]: "is" should be if

File: amethyst_core/src/transform/systems.rs [hunk 11]
```diff
@@ -31,87 +29,92 @@ impl TransformSystem {
...
-                if local.is_dirty() {
-                    self.dirty.insert(ent);
-                    let global = global_restrict.get_mut_unchecked(&mut global_entry);
-                    global.0 = local.matrix();
-                    local.flag(false);
-                }
+            for (_entity, local, global, _) in (&*entities, locals_flagged, &mut globals, !&parents).join() {
+                global.0 = local.matrix();
+                debug_assert!(global.0 == global.0, format!("Entity {:?} had NaN transform.", _entity));
             }
```
Comment (torkleyy) [inline amethyst_core/src/transform/systems.rs]: Should be `is_finite`

Comment (Aceeri) [reply amethyst_core/src/transform/systems.rs]: I don't think `[[f32; 4]; 4]` has methods implemented by default? `f32` has `is_finite`, but not this.

Comment (torkleyy) [reply amethyst_core/src/transform/systems.rs]: Ah right. I wonder if this is valid then? I'm not sure if the total equality check catches everything `is_finite` does

Comment (Aceeri) [reply amethyst_core/src/transform/systems.rs]: It doesn't check for Infinity, but I'm not sure if it matters if we allow that or not.

File: amethyst_core/src/transform/components/parent.rs (hunk 9 above)
Comment (omni-viral) [reply amethyst_core/src/transform/components/parent.rs]: Or `ChildOf`

File: amethyst_core/src/transform/systems.rs (hunk 11 above)
Comment (torkleyy) [reply amethyst_core/src/transform/systems.rs]: If that's the case this check doesn't make too much sense.

File: amethyst_core/src/transform/components/parent.rs (hunk 9 above)
Comment (Xaeroxe) [reply amethyst_core/src/transform/components/parent.rs]: This was addressed.

**Comment by Xaeroxe** [comment]
https://github.com/amethyst/amethyst/pull/403#discussion_r146096126

https://github.com/amethyst/amethyst/pull/403#discussion_r146096095

I'd just like to re-iterate @torkleyy 's expressed concerns.  Even if he's approved the PR I'm still requesting these changes be made.

**Comment by Rhuagh** [comment]
Also needs some fixes in the GLTF system (it uses Child atm, needs to be changed to Parent).
